### ISSUE TITLE
How to handle missing addons ?

### ISSUE THREAD

**Issue by Daandelange** [ofZach/pgElectron#12]
@kylemcdonald [mentions](https://github.com/openframeworks/projectGenerator/issues/5#issuecomment-132637138) 2 behaviours for handling the missing addons:
- is this an error, and it will not generate a project file? (probably the best behavior)
- is this a warning, and it automatically removes those addons? (dangerous)

Maybe the best thing to do is to put the missing addons back in the addons.make while generating the project anyways with the available addons ?

But.... does it make sense to generate it anyways, event if you don't have the needed addon ?
What is the OF philosophy behind this? Allow and learn from your errors I guess; so don't restrict the user ?

What if an addon vanishes from the internet? There should always be an option to remove a missing addon.

**Comment by ofZach** [comment]
I think we should give a warning but just generate the project without stopping them....  I can add some info to the output of the command line tool that flags the addon not being found.

**Comment by ofZach** [comment]
the warning can even be an alert prompt, like "these addons are missing: ..."    I think we should never call addons plugins anywhere, since we don't use that language elsewhere. 

I also feel like we generally should try to remove as much text as we can / simplfy the text if possible.  Things like missing addons can be a one time pop up, so you realize, "oof, I'm missing something".

Also, since updating may overwrite the addons.make file, we might want to be careful here -- I think the right thing is to pass them into the PG, so they get saved back into the addons.make (I will double test this) so that we don't "loose" the addons when we update a project, they just won't be in the project you generate.

**Comment by Daandelange** [comment]
Ok, that's pretty clear but then we also need to be able to remove a missing addon.
I'm thinking about a small red square with the number of addons missing. if you click on it, the list appears and you can eventually remove a missing addon from `addons.make`.

**Comment by ofZach** [comment]
I think even simpler is an overlay or alert saying 

"you are missing these addons:....  warning, if you update this project, your addons.make will get overwritten, these addons will not get added and this project may not work.  to remove these addons, edit addons.make.  If you add addons to your OF folder, restart this GUI and try again to import"  and then just don't worry about missing addons from a UI perspective.  I think it's best just to have an alert (or alert style overlay) that maybe happens once?

I feel like integrating missing addons into the UI is going to be maybe confusing and present alot of UI challenges since those things don't exist vs the actual addons which do exist.  I kind of don't think the PG gui is the right place to remove an addon that doesn't exist.

I also think in most cases, the alert would prompt someone to fix their problem. 

The other option is just don't do the import if the addons are missing -- maybe we can just throw up an alert so they have to fix it.  That will definitely prevent us having to deal with weird situations / UI paths :)

**Comment by Daandelange** [comment]
:+1: for the alert saying to manually remove it.

**Comment by ofZach** [comment]
the more I think about, the more I feel like we just want a helpful alert and just don't import projects that have missing addons.  We scan the addon folder only once at the start of the GUI, so even if they fix it, they still won't be able to import.  I think this can overlay can be helpful (with links to ofxAddons.com, etc) and also explain stuff like restarting to rescan the addon folder, etc.

**Comment by Daandelange** [comment]
I agree but I don't like the user being stuck just (don't import). I hate having to bypass such screens. :angel: 

_Side note: The addon folder is scanned every time the OFpath is updated, or it can very easily be triggered._

**Comment by ofZach** [comment]
I feel like it's easier then dealing with the nasty side effects of either (a) strange UI or (b) damaging a project by using the tool.  We don't have to deal with all edge cases, let's let the user fix the problem, either by adding the addon or modifying the addons.make, not use the tool to fix the problem (which seems hard to do).   I also think keeping it really simple for this release is important, then later, we can see what people get tripped up on.

**Comment by ofZach** [comment]
or if we want to let them muck up their project, we give them enough warnings -- and let them import after seeing the warning (and potentially loosing their missing addons from the project after updating).  Either way, I don't think we should try to be _smart_, I just think we should just try to give helpful warnings.,,,

**Comment by Daandelange** [comment]
Ok lets go for that, plus a "rescan addons" button on that warning screen to prevent relaunching the UI.

**Comment by ofZach** [comment]
ok sounds good

**Comment by kylemcdonald** [comment]
ok, the way i read this is:
- user has addons.make that includes ofxSyphon
- they try to generate/update a project file
- they get a warning because they don't have ofxSyphon
- the addons list in the GUI does not contain ofxSyphon
- they hit "generate" and a project file is generated without ofxSyphon
- the addons.make is not modified

if so, this sounds like the perfect solution.

if there is a "rescan" button on the warning, that would be good -- it would be best if it only appeared when necessary. it would be even better if we watched the addons folder for changes and automatically updated.

**Comment by ofZach** [comment]
I think still addons.make would get messed up unless we keep track of what addons are missing and "re-add" to the addons when we generate a project.  Personally, I think this is a little weird, like how do we signal that we are doing that?  I would just generally vote to not let someone import a project unless they have the addons present (but give them a useful warning) and then try to deal with being smarter after this version is done.

**Comment by ofZach** [comment]
this might be helpful for watching addons folder, that seems like a feature for down the line https://github.com/mikeal/watch

**Comment by Daandelange** [comment]
@kylemcdonald I felt like that too but why generate a project with missing addons ? It won't compile anyways, so "blocking" the user and explaining how to download the addon (or how to rm the addon from the project) also seems to be a good solution... (?)
As Zach says, it's less work to make it with the "blocking" warning screen (for OF 9.0) and «later, we can see what people get tripped up on» and rethink it. x)

**Comment by kylemcdonald** [comment]
Cool! Blocking them from doing anything was my preference originally anyway
:) but yes, the next feature to add then would be watching the adding
folder for updates.

**Comment by Daandelange** [comment]
I made a simple mockup for this. (Needs to be refined, specially the text messages)
